### PR TITLE
Add consistency check in the ATOMIC_SPECIES card block

### PR DESCRIPTION
**Pull request by asle85**
Validate species names in the `ATOMIC_SPECIES` card block. This will check that the atom names and the species names parsed from the pseudopotential file names are the same up to an optional underscore and/or optional digits.

### CODE REVIEW

File: qe_tools/parsers/_input_base.py [hunk 1]
```diff
@@ -147,6 +151,11 @@ def __init__(self, content, *, qe_version=None):
             Valid version strings are e.g. '6.5', '6.4.1', '6.4rc2'.
         :type qe_version: Optional[str]
 
+        :param validate_species_names: A boolean flag (default: True) to enable
+            the consistency check between atom names and species names inferred
+            from the pseudopotential file name.
+        :type validate_species_names: Optional[bool]
```
Comment (greschd) [inline qe_tools/parsers/_input_base.py:157]: The type should be `bool`, without the `Optional[]` qualifier (which indicates that `None` can also be passed).

File: qe_tools/parsers/_input_base.py [hunk 2]
```diff
@@ -679,6 +694,36 @@ def _parse_atomic_species(txt):
     :raises qe_tools.utils.exceptions.ParsingError: if there are issues
         parsing the input.
     """
+    def validate_species_name(atom_name, pseudo_file_name):
```
Comment (greschd) [inline qe_tools/parsers/_input_base.py:697]: Please add a leading underscore to the method name, to indicate that this is not part of the _public_ interface of the class.

File: qe_tools/parsers/_pw_input.py [hunk 3]
```diff
@@ -141,6 +145,11 @@ def __init__(self, content, *, qe_version=None):
             Valid version strings are e.g. '6.5', '6.4.1', '6.4rc2'.
         :type qe_version: Optional[str]
 
+        :param validate_species_names: A boolean flag (default: True) to enable
+            the consistency check between atom names and species names inferred
+            from the pseudopotential file name.
+        :type validate_species_names: Optional[bool]
```
Comment (greschd) [inline qe_tools/parsers/_pw_input.py:151]: Same as above (type `bool`).

**Review by greschd — CHANGES_REQUESTED** [review]
Thanks @asle85, this looks great! I just added some minor comments.

Would you mind also adding a test for this check? 

Here's an example for a test of a "failing" parsing:

https://github.com/aiidateam/qe-tools/blob/359579ee4a88fb6920270de39f49524e9365fda2/tests/test_parsers.py#L211-L214

Those labels refer to input files in the `tests/data` directory: https://github.com/aiidateam/qe-tools/tree/develop/tests/data

I think we should test also when `validate_species_names` is `False` - I think the easiest way to do that is adding it as a keyword argument to "singletest":

https://github.com/aiidateam/qe-tools/blob/359579ee4a88fb6920270de39f49524e9365fda2/tests/test_parsers.py#L111

**Comment by asle85** [comment]
Thanks @greschd, I've addressed your comments and added a test for the "failing" parsing.  

Regarding the test for the case `validate_species_names=False`, I get the expected `InputValidationError` when I try to create the reference file with `  ./test_parsers.py --write-ref non_matching_species`. 
One thing I can do is to create the reference json using the code of the `print_test_comparison` [function](https://github.com/aiidateam/qe-tools/blob/7a733be7c057c8df4a1505f0625594245fc4c42c/tests/test_parsers.py#L322). Or I could modify the function [here](https://github.com/aiidateam/qe-tools/blob/7a733be7c057c8df4a1505f0625594245fc4c42c/tests/test_parsers.py#L378) to add a `validate_species_names` parameter.

File: qe_tools/parsers/_input_base.py (hunk 1 above)
Comment (asle85) [reply qe_tools/parsers/_input_base.py]: right, done

File: qe_tools/parsers/_pw_input.py (hunk 3 above)
Comment (asle85) [reply qe_tools/parsers/_pw_input.py]: done :+1:

File: qe_tools/parsers/_input_base.py (hunk 2 above)
Comment (asle85) [reply qe_tools/parsers/_input_base.py]: done :+1:

**Review by greschd — DISMISSED** [review]
Thanks @asle85, all good to go!

**Comment by greschd** [comment]
> One thing I can do is to create the reference json using the code of the print_test_comparison function. Or I could modify the function here to add a validate_species_names parameter.

I think it would be ok to either add a flag, or just hard-code `validate_species_names=False` in the `print_test_comparison` function. Up to you if you want to add that to this PR.

**Comment by asle85** [comment]
I've just set `validate_species_names=False` [here](https://github.com/asle85/qe-tools/blob/394f17a36fe74534e123f5d97a9fef59c5af6b17/tests/test_parsers.py#L358), since if the test is supposed to fail there's no need for creating the json comparison.

**Review by greschd — APPROVED** [review]

**Comment by greschd** [comment]
> I've just set `validate_species_names=False` [here](https://github.com/asle85/qe-tools/blob/394f17a36fe74534e123f5d97a9fef59c5af6b17/tests/test_parsers.py#L358), since if the test is supposed to fail there's no need for creating the json comparison.

Perfect, thanks again @asle85!

**Pull request merged**